### PR TITLE
WebUI customisation update

### PR DESCRIPTION
**Pull request by mmattel**
Fixes: #743 (Add changes to the web service)

This adds the changes for web app added in ocis master.
Note that after content clarification with @fschade, we need to fix the web/readme in the ocis repo too.

Language review welcomed

Currently visible on [staging](https://doc.staging.owncloud.com/ocis/next/deployment/webui/webui-customisation.html)

No backport

### CODE REVIEW

File: modules/ROOT/pages/deployment/webui/webui-customisation.adoc [hunk 1]
```diff
@@ -23,3 +23,114 @@ When defined via the `WEB_UI_CONFIG_FILE` xref:{s-path}/web.adoc[environment var
 == Embedding Web
 
 Web can be consumed by another application in a stripped-down version called “Embed mode”. This mode is supposed to be used in the context of selecting or sharing resources. For more details refer to the developer documentation https://owncloud.dev/clients/web/embed-mode/[ownCloud Web / Embed Mode]. See the environment variables: `WEB_OPTION_MODE` and `WEB_OPTION_EMBED_TARGET` to configure the embedded mode.
+
+== Web Apps
+
+The administrator of the environment is capable of providing custom web applications to the users. This feature is useful for organizations that want to provide third party or custom apps to their users.
+
+It's important to note, that the feature is at the moment only capable providing static (js, mjs, e.g.) web applications and does not support injection of dynamic web applications (custom dynamic backends).
```
Comment (phil-davis) [inline modules/ROOT/pages/deployment/webui/webui-customisation.adoc:31]:
```suggestion
It's important to note, that the feature is at the moment only capable of providing static (js, mjs, e.g.) web applications and does not support injection of dynamic web applications (custom dynamic backends).
```

File: modules/ROOT/pages/deployment/webui/webui-customisation.adoc [hunk 2]
```diff
@@ -23,3 +23,114 @@ When defined via the `WEB_UI_CONFIG_FILE` xref:{s-path}/web.adoc[environment var
...
+
+=== Loading Applications
+
+Web applications are loaded, if added in the Infinite Scale source code, at build-time from `<ocis_repo>/services/web/assets/apps`. This cannot be manipulated at runtime.
+
+Additionally, the administrator can provide custom applications by storing them in the path defined the environment variable `WEB_ASSET_APPS_PATH`.
```
Comment (phil-davis) [inline modules/ROOT/pages/deployment/webui/webui-customisation.adoc:37]:
```suggestion
Additionally, the administrator can provide custom applications by storing them in the path defined by the environment variable `WEB_ASSET_APPS_PATH`.
```

File: modules/ROOT/pages/deployment/webui/webui-customisation.adoc [hunk 3]
```diff
@@ -23,3 +23,114 @@ When defined via the `WEB_UI_CONFIG_FILE` xref:{s-path}/web.adoc[environment var
...
+
+This environment variable defaults to the Infinite Scale xref:deployment/general/general-info.adoc#base-data-directory[base data directory] `$OCIS_BASE_DATA_PATH/web/assets/apps`, but can be redefined with any path set manually.
+
+The final list of available applications is composed of the build in and the custom applications provided by the administrator via `WEB_ASSET_APPS_PATH`.
```
Comment (phil-davis) [inline modules/ROOT/pages/deployment/webui/webui-customisation.adoc:41]:
```suggestion
The final list of available applications is composed of the built-in and the custom applications provided by the administrator via `WEB_ASSET_APPS_PATH`.
```

I think this is what you want to say.

File: modules/ROOT/pages/deployment/webui/webui-customisation.adoc [hunk 4]
```diff
@@ -23,3 +23,114 @@ When defined via the `WEB_UI_CONFIG_FILE` xref:{s-path}/web.adoc[environment var
...
+
+For example, if Infinite Scale would contain a build in extension named `image-viewer-dfx` and the administrator provides a custom application named `image-viewer-obj` via the `WEB_ASSET_APPS_PATH` directory, the user will be able to access both
```
Comment (phil-davis) [inline modules/ROOT/pages/deployment/webui/webui-customisation.adoc:43]:
```suggestion
For example, if Infinite Scale would contain a built-in extension named `image-viewer-dfx` and the administrator provides a custom application named `image-viewer-obj` via the `WEB_ASSET_APPS_PATH` directory, the user will be able to access both
```

File: modules/ROOT/pages/deployment/webui/webui-customisation.adoc [hunk 5]
```diff
@@ -23,3 +23,114 @@ When defined via the `WEB_UI_CONFIG_FILE` xref:{s-path}/web.adoc[environment var
...
+applications from the WebUI.
+
+=== Application Structure
+
+* Applications always have to follow a strict structure, which are:
```
Comment (phil-davis) [inline modules/ROOT/pages/deployment/webui/webui-customisation.adoc:48]:
```suggestion
* Applications always have to follow a strict structure, which is:
```

File: modules/ROOT/pages/deployment/webui/webui-customisation.adoc [hunk 6]
```diff
@@ -23,3 +23,114 @@ When defined via the `WEB_UI_CONFIG_FILE` xref:{s-path}/web.adoc[environment var
...
+NOTE: An application manifest should _never_ be changed manually, see xref:using-custom-assets[Using Custom Assets] for customisation.
+
+The `apps.yaml` file must contain a list of key-value pairs which gets merged with the `config` field. For example, if the `image-viewer-obj` application contains the following configuration:
+
+[source,json]
+----
+{
+  "entrypoint": "index.js",
+  "config": {
+    "maxWith": 1280,
```
Comment (phil-davis) [inline modules/ROOT/pages/deployment/webui/webui-customisation.adoc:77]:
```suggestion
    "maxWidth": 1280,
```

File: modules/ROOT/pages/deployment/webui/webui-customisation.adoc [hunk 7]
```diff
@@ -23,3 +23,114 @@ When defined via the `WEB_UI_CONFIG_FILE` xref:{s-path}/web.adoc[environment var
...
+
+[source,json]
+----
+{
+  "external_apps": [
+    {
+      "id": "image-viewer-obj",
+      "path": "index.js",
+      "config": {
+        "maxWith": 1280,
```
Comment (phil-davis) [inline modules/ROOT/pages/deployment/webui/webui-customisation.adoc:103]:
```suggestion
        "maxWidth": 1280,
```

File: modules/ROOT/pages/deployment/webui/webui-customisation.adoc [hunk 8]
```diff
@@ -23,3 +23,114 @@ When defined via the `WEB_UI_CONFIG_FILE` xref:{s-path}/web.adoc[environment var
...
+Consider the following: Infinite Scale is shipped with a default web app named `image-viewer-dfx` which contains a logo,
+but the administrator wants to provide a custom logo for that application.
+
+This can be achieved using the path defined via `WEB_ASSET_APPS_PATH` and adding a custom structure like `WEB_ASSET_APPS_PATH/image-viewer-dfx/`. Here you can add all custom assets to load like `logo.png`. On loading the web app, custom assets defined overwrite default ones.
+
+This also applies for the `manifest.json` file, if the administrator wants to provide a custom one.
+
+== Miscellaneous
+
+Please note that Infinite Scale, respectively the web service needs a restart to load new applications or changes to the `apps.yaml` file.
```
Comment (phil-davis) [inline modules/ROOT/pages/deployment/webui/webui-customisation.adoc:136]:
```suggestion
Please note that Infinite Scale, in particular the web service, needs a restart to load new applications or changes to the `apps.yaml` file.
```

**Review by dragonchaser — APPROVED** [review]

**Pull request merged**